// tdesign-vue-next/es/image/props.mjs
/**
 * tdesign v1.3.1
 * (c) 2023 tdesign
 * @license MIT
 */

var props = {
  alt: {
    type: String,
    "default": ""
  },
  error: {
    type: [String, Function]
  },
  fit: {
    type: String,
    "default": "fill",
    validator: function validator(val) {
      if (!val) return true;
      return ["contain", "cover", "fill", "none", "scale-down"].includes(val);
    }
  },
  gallery: Boolean,
  lazy: Boolean,
  loading: {
    type: [String, Function]
  },
  overlayContent: {
    type: [String, Function]
  },
  overlayTrigger: {
    type: String,
    "default": "always",
    validator: function validator(val) {
      if (!val) return true;
      return ["always", "hover"].includes(val);
    }
  },
  placeholder: {
    type: [String, Function]
  },
  position: {
    type: String,
    "default": "center"
  },
  shape: {
    type: String,
    "default": "square",
    validator: function validator(val) {
      if (!val) return true;
      return ["circle", "round", "square"].includes(val);
    }
  },
  src: {
    type: String,
    "default": ""
  },
  srcset: {
    type: Object
  },
  onError: Function,
  onLoad: Function
};

export { props as default };
